lib/internal/modules/esm/fetch_module.js

```diff
@@ -127,7 +148,7 @@ function fetchWithRedirects(parsed) {
         err.message = `Cannot find module '${parsed.href}', HTTP 404`;
         throw err;
       }
-      if (res.statusCode > 303 || res.statusCode < 200) {
+      if (res.statusCode < 200 || res.statusCode >= 400) {
```

Comment:
Should this also include 304, as in `(res.statusCode < 200 || res.statusCode === 304 || res.statusCode >= 400)`? @JakobJingleheimer? I feel like we should never get a 304 in Node because we’re not maintaining a cache in the way browsers are?

Also what is this error `'cannot redirect to non-network location'` referring to? If the status code is 500, this seems a weird error to throw.

Reply:
I would just switch this to the new `isRedirect()` helper

> I feel like we should never get a 304 in Node because we’re not maintaining a cache in the way browsers are?

We maintain _a_ cache, and I think there are designs (perhaps only in people's heads at the moment?) for a right-to-disk cache in future (at which point a 304 would be very valid). I'd say account for it now whilst we're thinking of it (especially because it's trivial) rather than get bitten by it later.

> Also what is this error 'cannot redirect to non-network location' referring to?

This is when a remote module tries to access a local module (eg `fs`), which is forbidden.

> If the status code is 500, this seems a weird error to throw.

Sorry, where are you seeing a `500`?

Reply:
Oh! Yes, indeed it should not throw disallowed network import for a 500.

@aduh95 I see you just approved; I'm think this should be consider a blocker (it introduces a bug). The rest can be addressed in a follow-up.

Reply:
Wrong error is not introduced in this PR, it's a part of original code.
What this PR in its current state changes here is behaviour for `303 < statusCode < 400` (for 307 and 308, only if no Location provided), importing body instead of throwing "non-network location" error. Which can be addressed in a follow-up, but is still in line with logic of old code (it did the same for `300~303`).

Reply:
> I see you just approved; I'm thinking this should be consider a blocker

Livia said it all, if there's a bug let's fix it in its own PR. This PR does a great job at adding support for 307 and 308, it wouldn't be fait to block it on a bug (or maybe not bug?) of the existing implementation.